pkg/devfile/adapters/kubernetes/component/adapter.go

```
@@ -571,6 +579,65 @@ func (a Adapter) generateDeploymentObjectMeta(deployment *appsv1.Deployment, lab
+		if !odolabels.IsCoreComponent(remoteK.GetLabels()) {
+			remoteK8sResources = append(remoteK8sResources, remoteK)
+		}
+	}
+	devfileK8sResources, err := devfile.GetKubernetesComponentsToPush(a.Devfile)
+	if err != nil {
+		return fmt.Errorf("unable to obtain devfile kubernetes resources: %w", err)
+	}
+
+	// convert all devfileK8sResources to unstructured data
+	var devfileK8sResourcesUnstructured []unstructured.Unstructured
+	for _, devfileK := range devfileK8sResources {
+		devfileKUnstructured, err := libdevfile.GetK8sComponentAsUnstructured(a.Devfile, devfileK.Name, a.Context, devfilefs.DefaultFs{})
+		if err != nil {
+			return fmt.Errorf("unable to obtain unstructured data: %w", err)
+		}
+		devfileK8sResourcesUnstructured = append(devfileK8sResourcesUnstructured, devfileKUnstructured)
+	}
+
+	var objectsToRemove []unstructured.Unstructured
+	for _, remoteK := range remoteK8sResources {
+		matchFound := false
+		for _, devfileK := range devfileK8sResourcesUnstructured {
+			if devfileK.GetKind() == remoteK.GetKind() && devfileK.GetName() == remoteK.GetName() && remoteK.GetNamespace() == a.kubeClient.GetCurrentNamespace() {
+				matchFound = true
+				break
+			}
+		}
+		if !matchFound {
+			objectsToRemove = append(objectsToRemove, remoteK)
+		}
+	}
+
+	// Delete the resources present on the cluster but not in the Devfile
+	for _, objectToRemove := range objectsToRemove {
+		log.Infof("Deleting Kubernetes resource: %s/%s", objectToRemove.GetKind(), objectToRemove.GetName())
+		gvr, err := a.kubeClient.GetGVRFromGVK(objectToRemove.GroupVersionKind())
+		if err != nil {
+			log.Warningf("Unable to get information about Kubernetes resource: %s/%s: %s", objectToRemove.GetKind(), objectToRemove.GetName(), err.Error())
+		}
+		err = a.kubeClient.DeleteDynamicResource(objectToRemove.GetName(), gvr, true)
+		if err != nil {
+			log.Warningf("Unable to delete Kubernetes resource: %s/%s: %s", objectToRemove.GetKind(), objectToRemove.GetName(), err.Error())
```

@feloy I am confused if we should just warn the error here or return the error? Wdyt?

---

You could check for the response: if the error is NotFound, you could ignore it, elsewhere you could return the error